File: airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretPersistence.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.config.persistence.split_secrets;

import io.airbyte.config.Configs;
import io.airbyte.db.Database;
import java.util.Optional;
import javax.annotation.Nullable;
import org.jooq.DSLContext;

/**
 * Provides the ability to read and write secrets to a backing store. Assumes that secret payloads
 * are always strings. See {@link SecretCoordinate} for more information on how secrets are
 * identified.
 */
@SuppressWarnings("PMD.MissingOverride")
public interface SecretPersistence extends ReadOnlySecretPersistence {

  /**
   * Performs any initialization prior to utilization of the persistence object. This exists to make
   * it possible to create instances within a dependency management framework, where any
   * initialization logic should not be present in a constructor.
   *
   * @throws Exception if unable to perform the initialization.
   */
  default void initialize() throws Exception {}

  Optional<String> read(final SecretCoordinate coordinate);

  void write(final SecretCoordinate coordinate, final String payload);

  /**
   * Get secrets persistence for long-lived secrets.
   *
   * @param dslContext db context
   * @param configs application configurations
   * @return secrets persistence for long-lived secrets
   */
  static Optional<SecretPersistence> getLongLived(final @Nullable DSLContext dslContext, final Configs configs) {
    switch (configs.getSecretPersistenceType()) {
      case TESTING_CONFIG_DB_TABLE -> {
        final Database configDatabase = new Database(dslContext);
        return Optional.of(new LocalTestingSecretPersistence(configDatabase));
      }
      case GOOGLE_SECRET_MANAGER -> {
        return Optional.of(GoogleSecretManagerPersistence.getLongLived(configs.getSecretStoreGcpProjectId(), configs.getSecretStoreGcpCredentials()));
      }
      case VAULT -> {
        return Optional.of(new VaultSecretPersistence(configs.getVaultAddress(), configs.getVaultPrefix(), configs.getVaultToken()));
      }
      case AWS_SECRET_MANAGER -> {
        return Optional.of(new AWSSecretManagerPersistence(configs.getAwsAccessKey(), configs.getAwsSecretAccessKey()));
      }
      default -> {
        return Optional.empty();
      }
    }
  }

  /**
   * Get secrets hydrator.
   *
   * @param dslContext db context
   * @param configs application configurations
   * @return secrets hydrator
   */
  static SecretsHydrator getSecretsHydrator(final @Nullable DSLContext dslContext, final Configs configs) {
    final var persistence = getLongLived(dslContext, configs);

    if (persistence.isPresent()) {
      return new RealSecretsHydrator(persistence.get());
    } else {
      return new NoOpSecretsHydrator();
    }
  }

  /**
   * Get secrets persistence for ephemeral secrets.
   *
   * @param dslContext db context
   * @param configs application configurations
   * @return secrets persistence for ephemeral secrets
   */
  static Optional<SecretPersistence> getEphemeral(final DSLContext dslContext, final Configs configs) {
    switch (configs.getSecretPersistenceType()) {
      case TESTING_CONFIG_DB_TABLE -> {
        final Database configDatabase = new Database(dslContext);
        return Optional.of(new LocalTestingSecretPersistence(configDatabase));
      }
      case GOOGLE_SECRET_MANAGER -> {
        return Optional.of(GoogleSecretManagerPersistence.getEphemeral(configs.getSecretStoreGcpProjectId(), configs.getSecretStoreGcpCredentials()));
      }
      case VAULT -> {
        return Optional.of(new VaultSecretPersistence(configs.getVaultAddress(), configs.getVaultPrefix(), configs.getVaultToken()));
      }
      case AWS_SECRET_MANAGER -> {
        return Optional.of(new AWSSecretManagerPersistence(configs.getAwsAccessKey(), configs.getAwsSecretAccessKey()));
      }
      default -> {
        return Optional.empty();
      }
    }
  }

}

File: airbyte-config/config-secrets/src/main/kotlin/secrets/persistence/SecretPersistence.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.config.secrets.persistence

import io.airbyte.config.secrets.SecretCoordinate
import java.time.Instant

/**
 * Provides a read-only interface to a backing secrets store similar to [SecretPersistence].
 * In practice, the functionality should be provided by a [SecretPersistence.read] function.
 */
fun interface ReadOnlySecretPersistence {
  fun read(coordinate: SecretCoordinate): String
}

/**
 * Provides the ability to read and write secrets to a backing store. Assumes that secret payloads
 * are always strings. See {@link SecretCoordinate} for more information on how secrets are
 * identified.
 */
interface SecretPersistence : ReadOnlySecretPersistence {
  /**
   * Performs any initialization prior to utilization of the persistence object. This exists to make
   * it possible to create instances within a dependency management framework, where any
   * initialization logic should not be present in a constructor.
   *
   * @throws Exception if unable to perform the initialization.
   */
  @Throws(Exception::class)
  fun initialize() {}

  override fun read(coordinate: SecretCoordinate): String

  fun write(
    coordinate: SecretCoordinate,
    payload: String,
  )

  fun writeWithExpiry(
    coordinate: SecretCoordinate,
    payload: String,
    expiry: Instant? = null,
  ) {
    // Default implementation does not support expiry.
    write(coordinate, payload)
  }

  fun delete(coordinate: SecretCoordinate)

  fun disable(coordinate: SecretCoordinate) {
    println("secret persistence has not implemented disable.")
  }
}
